Add Sinkhorn PIT loss

tachi-hi opened the pull request:
### asteroid/engine/system.py 
- A change was made to pass the epoch number to the PIT wrapper.

### asteroid/losses/pit_wrapper.py
- Add Sinkhorn iteration-based permutation search.
    - Some parameters are still hard coded.
- Possibility of stopping on ASSERT during the validation, which is painful if it occurs in a very heavy training. We may need to devise a way to separate valid and test by flags. Replacing it with the Hungarian algorithm is another option.

### tests
- Check if Sinkhorn and Hungarian return almost the same results for 2 <= N <=9.
- Shape check.
- Check if Sinkhorn-based reordering is correct.

mpariente commented:
Thanks a lot for the implementation, it's great !! 
Is this ready for review in your opinion? Should I give you feedback?

tachi-hi commented:
I'd be grateful for all kinds of feedback!

The test was successful in my local environment, but the CI here failed.
(Failure in quantizing a soft permutation matrix into a hard permutation matrix.)
It is probably due to a difference in the seeds of random numbers.

In light of the CI results, it may be better to use the strict Hungarian algorithm than 
to quantize a soft permutation matrix, during the valid/test(`reorder=True`) stage. 
(There is also an option to leave both and select one of them with a flag, 
but it may be more complicated than necessary).

mpariente commented on asteroid/engine/system.py:
```diff
@@ -91,7 +93,10 @@ def common_step(self, batch, batch_nb, train=True):
         inputs, targets = batch
         est_targets = self(inputs)
-        loss = self.loss_func(est_targets, targets)
+        if isinstance(self.loss_func, PITLossWrapper):
+            loss = self.loss_func(est_targets, targets, epoch=self.current_epoch)
+        else:
+            loss = self.loss_func(est_targets, targets)
```
I don't think we'll accept that. 
Otherwise, we'll never stop growing the codebase for specific use cases.

tachi-hi replied:
OK!
In the modified version (in another branch), I've tried using a callback class that updates the `beta` at the beginning of each epoch.

mpariente replied:
Cool !
Feel free to open a PR with the other branch :smiley:

mpariente reviewed (commented):
Overall, it looks good, thanks ! 

I think we should rather implement `SinkPITLossWrapper` instead of adding yet another mode to `PITLossWrapper`. 
SinkPIT is actually quite different because it retains more informations than the other methods, and it would make sense to make it a different object altogether. Could you implement it as a separate class please? That would be great !

Also, there would be less constraint on it, and it would make experimentation easier. I'll have a look at the tests when it's a separate class.

mpariente commented:
I see a TODO in the docstring, but otherwise this seems ready. Should it be reviewed?

tachi-hi commented:
I think I'm ready to ask for a code review.
- Implemented a `SinkPITLossWrapper` class
- Epoch number is passed to `SinkPITLossWrapper` through an inheritance of `pl.callbacks.Callback`.
- I filled in the TODO in the docstring with a snippet.
- Added an option whether to use the Hungarian algorithm for the validation.
- I did an experiment replacing the PIT in an `egs` sample with SinkPIT. It seems to be working (the loss is decreasing). But I haven't written a test for this part.

mpariente commented on asteroid/losses/sinkpit_wrapper.py:
```diff
@@ -0,0 +1,211 @@
+        >>> trainer.fit(system)
+    """
+
+    def __init__(self, loss_func, n_iter=200, hungarian_validation=True, *args, **kwargs):
```
`args` and `kwargs` are unused, why are they here?

mpariente commented:
Did you untick `Allow edits from maintainers`? 
I tried to commit to the PR but couldn't.

mpariente reviewed (commented):
This is mostly ready for merging, thanks. 

I wonder, would it make sense to make `PITBase` class with all the static methods, and inherit both wrappers from it? It would be more pythonic IMO. @jonashaag?

After we decide about, I'd like to change a few details and this is ready. 
The docstrings are really pretty, thanks for the extra mile !

mpariente commented:
> Did you untick `Allow edits from maintainers`?
> I tried to commit to the PR but couldn't.

My bad, I made a silly mistake. I could push now

jonashaag commented on asteroid/losses/sinkpit_wrapper.py:
```diff
@@ -0,0 +1,192 @@
+        and http://arxiv.org/abs/2010.11871
+
+    Examples
+        >>> import torch
```
This is not something that we have to fix in this PR, but still: Since none of these lines actually show the return values of any of the expressions, I think we can do away with all of the `>>>`. (Btw they aren't "correct" since in the multiline expressions below the lines should start with `...`.) What do you think @mpariente?

mpariente replied:
You mean remove them right? 
I don't know how that renders in the docs without `>>>`, I'd have to check.

jonashaag replied:
This is Sphinx, right?

You can either do

```
Examples::

	from ... import ...
```

which renders as `Examples: ...`

or

```
Examples ::

	from ... import ...
```

which renders as `Examples ...` (no `:`)

mpariente replied:
I don't know sphinx that well, it's always a mistery why some things work and others don't.

mpariente replied:
If you have a pointer to a succinct and good guide to Sphinx, I'll take it. 
I have a unfinished document called `writing_good_docstrings`, I still don't know how to :roll_eyes:

jonashaag replied:
Haha yeah I was thinking about that too. I also don't have a clue. Maybe sometime we can spend some time researching this and coming up with a guide. Sometime...

jonashaag replied:
Maybe just look at PyTorch or NumPy docstrings

jonashaag commented on tests/losses/sinkpit_wrapper_test.py:
```diff
@@ -0,0 +1,150 @@
+@pytest.mark.parametrize("n_src", [2, 3, 4])
+@pytest.mark.parametrize("beta,n_iter", [(100.0, 2000)])
+@pytest.mark.parametrize(
+    "function_triplet",
```
Maybe import this from the other loss tests instead of redefining it here?

mpariente replied:
How to import locally in tests?

jonashaag replied:
Well you can just do `from .loss_functions_test import ...` but it only works if you properly package all the tests (`__init__.py` files etc). I'll resolve this as "not worth the effort" :) A little duplication is okay

mpariente replied:
Agree. Plus, this file is not supposed to test all the loss functions (it doesn't hurt though).

jonashaag commented on tests/losses/sinkpit_wrapper_test.py:
```diff
@@ -0,0 +1,150 @@
+    assert_allclose(mean_loss_sinkhorn, mean_loss_hungarian)
+
+
+class testCallback(pl.callbacks.Callback):
```
This should have an uppercase T: `TestCallback`. I wonder why the linters don't complain about it.

mpariente replied:
Also don't know. 
Useless linter ^^

jonashaag commented:
> I wonder, would it make sense to make `PITBase` class with all the static methods, and inherit both wrappers from it? It would be more pythonic IMO. @jonashaag?

What's the reason for having these in the class in the first place? Convenience? Or so that subclasses can override them?

mpariente commented:
Convenience.
I use the static method from `PITLossWrapper` in the end.

mpariente commented on tests/losses/sinkpit_wrapper_test.py:
```diff
@@ -0,0 +1,150 @@
+
+    # compute loss by sinkhorn
+    loss_sinkhorn.beta = beta
+    mean_loss_sinkhorn = loss_hungarian(est_targets, targets, return_est=False)
```
That's not a test, it's `loss_hungarian`.. On it.

jonashaag commented on tests/losses/sinkpit_wrapper_test.py:
```diff
@@ -65,7 +65,7 @@ def test_wrapper(batch_size, n_src, time):
         [pairwise_mse, singlesrc_mse, multisrc_mse],
     ],
 )
-def test_proximity_sinkhorn_hungrian(batch_size, n_src, beta, n_iter, function_triplet):
```
![grafik](https://user-images.githubusercontent.com/175722/98410308-b59a1000-2074-11eb-91f5-482bca611219.png)

HUNGRIAN
